### DISCUSSION TITLE
Denormalized Context by Default

### DISCUSSION THREAD

**Discussion by csirmazbendeguz** [AmbitionEng/django-pghistory#177] [Ideas]
First of all, let me say this library is amazing. I have read through the docs today and I have a suggestion.

I was surprised to see `pgh_context` is a `ForeignKey` by default.

If I understand correctly, the only benefit of storing the context in a table over a `JSONField` is that since context objects can be shared between events, they can be updated after an event is created. I don't know if there's a good use case for this though?

I think [denormalized contexts](https://django-pghistory.readthedocs.io/en/3.5.0/event_models/?h=pgh_obj#denormalizing-context) would make more sense as the default setting. WDYT?

**Comment by wesleykendall** [comment]
Thank you for the suggestion!

There are a few reasons it isn't the default:

1. Denormalized context was introduced in later versions, so making it the default would require an API break and communication with users that upgrade. While this is part of life for upgrading libraries, I've tried to avoid API breaks when possible
2. Say that a user is tracking context in an API request and makes 1000 updates. If denorming context, you'll have 1000 different copies of (almost always) the exact same JSON across all of your event rows instead of all of those event rows pointing to the same context. Similarly, if your app adds context midway during the request, only the later events after that additional context will reference the new context. This may be desirable or not depending on the situation.

One of the original reasons for supporting denormed context was to also better support partitioning of the event tables. This is a pretty advanced edge case.

Given that denormed context can be turned on globally for everything with one setting, I just opted to keep it as something people opt in to if they are aware of the additional space requirements and other subtle differences

**Reply by csirmazbendeguz** [reply]
I see, thanks for the quick reply and for your work on this library!
